File: android/notion/src/main/java/com/prince/notion/NotionClient.java
package com.prince.notion;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Thin REST wrapper for the two Notion endpoints this module needs:
 * <ul>
 *   <li>{@code POST /v1/search} — fetch granted top-level pages so the user can pick a
 *       default parent during connect.</li>
 *   <li>{@code POST /v1/pages} — create a child page under the chosen parent. The body
 *       is built from the structured output of {@link NotionLlmBridge}.</li>
 * </ul>
 *
 * <p>All calls run on a single-threaded background executor so the IME thread is never
 * blocked. Callbacks are invoked off the main thread — call sites that touch UI must
 * post to a Handler themselves.
 */
public final class NotionClient {

    private static final String TAG = "NotionClient";
    private static final String API_VERSION = "2022-06-28";
    private static final String BASE = "https://api.notion.com";

    private static final ExecutorService EXEC = Executors.newSingleThreadExecutor();

    public interface PageCallback {
        void onSuccess(String pageId, String pageUrl);
        void onError(String reason);
    }

    public interface SearchCallback {
        void onResults(List<Page> pages);
        void onError(String reason);
    }

    public static final class Page {
        public final String id;
        public final String title;
        public Page(String id, String title) { this.id = id; this.title = title; }
    }

    private final String accessToken;

    public NotionClient(String accessToken) {
        this.accessToken = accessToken;
    }

    public void searchPages(SearchCallback cb) {
        EXEC.execute(() -> {
            HttpURLConnection conn = null;
            try {
                JSONObject body = new JSONObject()
                        .put("filter", new JSONObject()
                                .put("value", "page")
                                .put("property", "object"));
                conn = openPost("/v1/search");
                writeBody(conn, body);
                int code = conn.getResponseCode();
                String resp = readAll(code < 400 ? conn.getInputStream() : conn.getErrorStream());
                if (code >= 400) { cb.onError("search_http_" + code + ": " + resp); return; }

                JSONObject json = new JSONObject(resp);
                JSONArray arr = json.optJSONArray("results");
                java.util.List<Page> out = new java.util.ArrayList<>();
                if (arr != null) {
                    for (int i = 0; i < arr.length(); i++) {
                        JSONObject p = arr.getJSONObject(i);
                        String id = p.optString("id");
                        String title = extractTitle(p);
                        if (id != null && !id.isEmpty()) {
                            out.add(new Page(id, title == null || title.isEmpty() ? "(untitled)" : title));
                        }
                    }
                }
                cb.onResults(out);
            } catch (Exception e) {
                Log.w(TAG, "search failed", e);
                cb.onError(e.getClass().getSimpleName() + ": " + e.getMessage());
            } finally {
                if (conn != null) conn.disconnect();
            }
        });
    }

    public void createPage(String parentPageId, String title, JSONArray blocks, PageCallback cb) {
        EXEC.execute(() -> {
            HttpURLConnection conn = null;
            try {
                JSONObject body = new JSONObject()
                        .put("parent", new JSONObject().put("page_id", parentPageId))
                        .put("properties", new JSONObject()
                                .put("title", new JSONArray()
                                        .put(textObject(title == null ? "Untitled" : title))))
                        .put("children", blocks == null ? new JSONArray() : blocks);
                conn = openPost("/v1/pages");
                writeBody(conn, body);
                int code = conn.getResponseCode();
                String resp = readAll(code < 400 ? conn.getInputStream() : conn.getErrorStream());
                if (code >= 400) { cb.onError("create_http_" + code + ": " + resp); return; }

                JSONObject json = new JSONObject(resp);
                String id = json.optString("id");
                String url = json.optString("url");
                if (id == null || id.isEmpty()) { cb.onError("no_page_id"); return; }
                cb.onSuccess(id, url);
            } catch (Exception e) {
                Log.w(TAG, "createPage failed", e);
                cb.onError(e.getClass().getSimpleName() + ": " + e.getMessage());
            } finally {
                if (conn != null) conn.disconnect();
            }
        });
    }

    // --- helpers ----------------------------------------------------------

    private HttpURLConnection openPost(String path) throws java.io.IOException {
        HttpURLConnection conn = (HttpURLConnection) new URL(BASE + path).openConnection();
        conn.setRequestMethod("POST");
        conn.setDoOutput(true);
        conn.setRequestProperty("Authorization", "Bearer " + accessToken);
        conn.setRequestProperty("Notion-Version", API_VERSION);
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setRequestProperty("Accept", "application/json");
        conn.setConnectTimeout(15_000);
        conn.setReadTimeout(20_000);
        return conn;
    }

    private void writeBody(HttpURLConnection conn, JSONObject body) throws java.io.IOException {
        try (OutputStream os = conn.getOutputStream()) {
            os.write(body.toString().getBytes(StandardCharsets.UTF_8));
        }
    }

    private static String readAll(InputStream is) throws java.io.IOException {
        if (is == null) return "";
        StringBuilder sb = new StringBuilder();
        try (BufferedReader r = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8))) {
            String line;
            while ((line = r.readLine()) != null) sb.append(line);
        }
        return sb.toString();
    }

    private static String extractTitle(JSONObject page) {
        JSONObject props = page.optJSONObject("properties");
        if (props == null) return null;
        // Walk properties to find the one whose type is "title".
        java.util.Iterator<String> keys = props.keys();
        while (keys.hasNext()) {
            JSONObject p = props.optJSONObject(keys.next());
            if (p == null) continue;
            if (!"title".equals(p.optString("type"))) continue;
            JSONArray rich = p.optJSONArray("title");
            if (rich == null || rich.length() == 0) return null;
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < rich.length(); i++) {
                JSONObject seg = rich.optJSONObject(i);
                if (seg != null) sb.append(seg.optString("plain_text", ""));
            }
            return sb.toString();
        }
        return null;
    }

    /** Build a Notion {@code rich_text} object wrapping plain text. */
    static JSONObject textObject(String text) {
        try {
            return new JSONObject()
                    .put("type", "text")
                    .put("text", new JSONObject().put("content", text == null ? "" : text));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: ios/TurtleKeyboardExtension/Integration/Notion/NotionClient.swift
import Foundation

/// Thin REST wrapper for the two Notion endpoints the module uses:
/// - `POST /v1/search` — list top-level pages so the user can pick a parent
/// - `POST /v1/pages`  — create a child page under the chosen parent
///
/// Direct port of Android's `NotionClient`. Token is the caller's problem.
enum NotionClient {

    private static let apiVersion = "2022-06-28"
    private static let base = "https://api.notion.com"
    private static let maxResponseBytes = 2_097_152
    private static let session: URLSession = {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 20
        config.timeoutIntervalForResource = 40
        config.urlCache = nil
        config.requestCachePolicy = .reloadIgnoringLocalCacheData
        config.waitsForConnectivity = false
        config.httpMaximumConnectionsPerHost = 2
        return URLSession(configuration: config)
    }()

    static func cancelAllRequests() { session.getAllTasks { $0.forEach { $0.cancel() } } }

    enum NotionError: Error, LocalizedError {
        case http(Int, String)
        case decode(String)

        var errorDescription: String? {
            switch self {
            case .http(let c, let m): return "HTTP \(c): \(m)"
            case .decode(let m):      return "Decode: \(m)"
            }
        }
    }

    /// Fetches every page the integration has been granted access to.
    static func searchPages(accessToken: String) async throws -> [NotionPage] {
        let body: [String: Any] = [
            "filter": ["value": "page", "property": "object"],
        ]
        let resp = try await post(path: "/v1/search", accessToken: accessToken, body: body)
        guard let arr = resp["results"] as? [[String: Any]] else { return [] }
        var out: [NotionPage] = []
        out.reserveCapacity(arr.count)
        for p in arr {
            guard let id = p["id"] as? String, !id.isEmpty else { continue }
            let title = extractTitle(page: p) ?? ""
            out.append(NotionPage(id: id, title: title.isEmpty ? "(untitled)" : title))
        }
        return out
    }

    /// Creates a child page under `parentPageId` titled `title`, with the
    /// given block array as children. `blocks` is the same shape the LLM
    /// bridge produces.
    static func createPage(
        accessToken: String,
        parentPageId: String,
        title: String,
        blocks: [[String: Any]]
    ) async throws -> (pageId: String, pageURL: String?) {
        let body: [String: Any] = [
            "parent": ["page_id": parentPageId],
            "properties": [
                "title": [richText(title.isEmpty ? "Untitled" : title)],
            ],
            "children": blocks,
        ]
        let resp = try await post(path: "/v1/pages", accessToken: accessToken, body: body)
        guard let id = resp["id"] as? String, !id.isEmpty else {
            throw NotionError.decode("no page id")
        }
        let url = resp["url"] as? String
        return (id, url)
    }

    // MARK: - Block helpers

    /// Build a Notion block JSON from a (type, text, checked) triple.
    /// `type` is one of "heading_2", "paragraph", "to_do".
    static func buildBlock(type: String, text: String, checked: Bool) -> [String: Any] {
        var inner: [String: Any] = [
            "rich_text": [richText(text)],
        ]
        if type == "to_do" {
            inner["checked"] = checked
        }
        return [
            "object": "block",
            "type": type,
            type: inner,
        ]
    }

    /// Build a Notion `rich_text` object wrapping plain text.
    static func richText(_ text: String) -> [String: Any] {
        [
            "type": "text",
            "text": ["content": text],
        ]
    }

    // MARK: - Internals

    private static func post(
        path: String,
        accessToken: String,
        body: [String: Any]
    ) async throws -> [String: Any] {
        guard let url = URL(string: base + path) else {
            throw NotionError.decode("bad URL")
        }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        req.setValue(apiVersion, forHTTPHeaderField: "Notion-Version")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        req.timeoutInterval = 20

        let (data, resp) = try await session.data(for: req)
        guard data.count <= maxResponseBytes else { throw NotionError.decode("response too large") }
        guard let http = resp as? HTTPURLResponse else {
            throw NotionError.http(0, "no response")
        }
        guard (200..<300).contains(http.statusCode) else {
            let msg = String(data: data, encoding: .utf8) ?? ""
            throw NotionError.http(http.statusCode, msg)
        }
        guard !data.isEmpty,
              let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return [:] }
        return json
    }

    private static func extractTitle(page: [String: Any]) -> String? {
        guard let props = page["properties"] as? [String: Any] else { return nil }
        for (_, value) in props {
            guard let p = value as? [String: Any],
                  let type = p["type"] as? String, type == "title",
                  let rich = p["title"] as? [[String: Any]]
            else { continue }
            return rich.compactMap { $0["plain_text"] as? String }.joined()
        }
        return nil
    }
}
